File: tinyagent/agent_types.py
"""Type definitions for the agent loop."""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator, Awaitable, Callable
from dataclasses import dataclass, field
from enum import Enum
from typing import Literal, Protocol, TypeAlias, TypeVar, Union, cast

from pydantic import BaseModel, ConfigDict, Field
from typing_extensions import TypeAliasType

# ------------------------------
# JSON-ish helper types
# ------------------------------

JsonPrimitive: TypeAlias = str | int | float | bool | None
JsonValue = TypeAliasType(
    "JsonValue",
    "JsonPrimitive | list[JsonValue] | dict[str, JsonValue]",
)
JsonObject: TypeAlias = dict[str, JsonValue]

ZERO_USAGE: JsonObject = {
    "input": 0,
    "output": 0,
    "cache_read": 0,
    "cache_write": 0,
    "total_tokens": 0,
    "cost": {
        "input": 0.0,
        "output": 0.0,
        "cache_read": 0.0,
        "cache_write": 0.0,
        "total": 0.0,
    },
}


# ------------------------------
# Core message/content types
# ------------------------------


class _AgentBaseModel(BaseModel):
    """Shared model configuration for migrated message/state models."""

    model_config = ConfigDict(arbitrary_types_allowed=True, extra="allow")


class ThinkingBudgets(_AgentBaseModel):
    """Token budgets for thinking/reasoning."""

    thinking_budget: int | None = None
    max_tokens: int | None = None


TResult = TypeVar("TResult")

MaybeAwaitable: TypeAlias = TResult | Awaitable[TResult]


class ThinkingLevel(str, Enum):
    """Thinking/reasoning level for models that support it."""

    OFF = "off"
    MINIMAL = "minimal"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    XHIGH = "xhigh"


class CacheControl(_AgentBaseModel):
    """Cache control directive for Anthropic prompt caching."""

    type: str | None = None


class TextContent(_AgentBaseModel):
    """Text content block."""

    type: Literal["text"] = "text"
    text: str | None = None
    text_signature: str | None = None
    cache_control: CacheControl | None = None


class ImageContent(_AgentBaseModel):
    """Image content block."""

    type: Literal["image"] = "image"
    url: str | None = None
    mime_type: str | None = None


class ThinkingContent(_AgentBaseModel):
    """Thinking content block."""

    type: Literal["thinking"] = "thinking"
    thinking: str | None = None
    thinking_signature: str | None = None
    cache_control: CacheControl | None = None


class ToolCallContent(_AgentBaseModel):
    """Tool call content block."""

    type: Literal["tool_call"] = "tool_call"
    id: str | None = None
    name: str | None = None
    arguments: JsonObject = Field(default_factory=dict)
    partial_json: str | None = None


ToolCall: TypeAlias = ToolCallContent

AssistantContent: TypeAlias = TextContent | ThinkingContent | ToolCallContent


class UserMessage(_AgentBaseModel):
    """User message for LLM."""

    role: Literal["user"] = "user"
    content: list[TextContent | ImageContent] = Field(default_factory=list)
    timestamp: int | None = None


StopReason: TypeAlias = Literal[
    "complete",
    "error",
    "aborted",
    "tool_calls",
    "stop",
    "length",
    "tool_use",
]

STOP_REASONS: frozenset[StopReason] = frozenset(
    {
        "complete",
        "error",
        "aborted",
        "tool_calls",
        "stop",
        "length",
        "tool_use",
    }
)


class AssistantMessage(_AgentBaseModel):
    """Assistant message from LLM."""

    role: Literal["assistant"] = "assistant"
    content: list[AssistantContent | None] = Field(default_factory=list)
    stop_reason: StopReason | None = None
    timestamp: int | None = None
    api: str | None = None
    provider: str | None = None
    model: str | None = None
    usage: JsonObject | None = None
    error_message: str | None = None


class ToolResultMessage(_AgentBaseModel):
    """Tool result message."""

    role: Literal["tool_result"] = "tool_result"
    tool_call_id: str | None = None
    tool_name: str | None = None
    content: list[TextContent | ImageContent] = Field(default_factory=list)
    details: JsonObject = Field(default_factory=dict)
    is_error: bool = False
    timestamp: int | None = None


Message = Union[UserMessage, AssistantMessage, ToolResultMessage]


class CustomAgentMessage(_AgentBaseModel):
    """Base class for custom agent messages."""

    role: str = ""
    timestamp: int | None = None


AgentMessage = Union[Message, CustomAgentMessage]


ConvertToLlmFn: TypeAlias = Callable[[list[AgentMessage]], MaybeAwaitable[list[Message]]]
TransformContextFn: TypeAlias = Callable[
    [list[AgentMessage], asyncio.Event | None],
    Awaitable[list[AgentMessage]],
]
ApiKeyResolver: TypeAlias = Callable[[str], MaybeAwaitable[str | None]]
AgentMessageProvider: TypeAlias = Callable[[], Awaitable[list[AgentMessage]]]


# ------------------------------
# Tool types
# ------------------------------


@dataclass
class AgentToolResult:
    """Result from executing a tool."""

    content: list[TextContent | ImageContent] = field(default_factory=list)
    details: JsonObject = field(default_factory=dict)


AgentToolUpdateCallback = Callable[[AgentToolResult], None]


@dataclass
class Tool:
    """Tool definition."""

    name: str = ""
    description: str = ""
    parameters: JsonObject = field(default_factory=dict)


@dataclass
class AgentTool(Tool):
    """Agent tool with execute function."""

    label: str = ""
    execute: Callable[..., Awaitable[AgentToolResult]] | None = None


# ------------------------------
# Context/model types
# ------------------------------


@dataclass
class Context:
    """Context for LLM calls."""

    system_prompt: str = ""
    messages: list[Message] = field(default_factory=list)
    tools: list[AgentTool] | None = None


@dataclass
class AgentContext:
    """Agent context with AgentMessage types."""

    system_prompt: str = ""
    messages: list[AgentMessage] = field(default_factory=list)
    tools: list[AgentTool] | None = None


@dataclass
class Model:
    """Model configuration."""

    provider: str = ""
    id: str = ""  # Model identifier (e.g., "gpt-4", "claude-3.5-sonnet")
    api: str = ""  # API type (e.g., "openai", "anthropic", "openrouter")
    thinking_level: ThinkingLevel = ThinkingLevel.OFF


class SimpleStreamOptions(_AgentBaseModel):
    """Standard stream options passed to providers."""

    api_key: str | None = None
    temperature: float | None = None
    max_tokens: int | None = None
    signal: asyncio.Event | None = None


StreamFn: TypeAlias = Callable[[Model, Context, SimpleStreamOptions], Awaitable["StreamResponse"]]


class AssistantMessageEvent(_AgentBaseModel):
    """Event during assistant message streaming."""

    type: (
        Literal[
            "start",
            "text_start",
            "text_delta",
            "text_end",
            "thinking_start",
            "thinking_delta",
            "thinking_end",
            "tool_call_start",
            "tool_call_delta",
            "tool_call_end",
            "done",
            "error",
        ]
        | None
    ) = None
    partial: AssistantMessage | None = None
    content_index: int | None = None
    delta: str | None = None
    content: str | TextContent | ThinkingContent | ToolCallContent | None = None
    tool_call: ToolCallContent | None = None
    reason: str | None = None
    message: AssistantMessage | None = None
    error: AssistantMessage | str | None = None


STREAM_UPDATE_EVENTS: frozenset[str] = frozenset(
    {
        "text_start",
        "text_delta",
        "text_end",
        "thinking_start",
        "thinking_delta",
        "thinking_end",
        "tool_call_start",
        "tool_call_delta",
        "tool_call_end",
    }
)


class StreamResponse(Protocol):
    """Response from streaming."""

    def result(self) -> Awaitable[AssistantMessage]: ...

    def __aiter__(self) -> AsyncIterator[AssistantMessageEvent]: ...

    async def __anext__(self) -> AssistantMessageEvent: ...


# ------------------------------
# Agent event types
# ------------------------------


@dataclass
class AgentStartEvent:
    type: Literal["agent_start"] = "agent_start"


@dataclass
class AgentEndEvent:
    type: Literal["agent_end"] = "agent_end"
    messages: list[AgentMessage] = field(default_factory=list)


@dataclass
class TurnStartEvent:
    type: Literal["turn_start"] = "turn_start"


@dataclass
class TurnEndEvent:
    type: Literal["turn_end"] = "turn_end"
    message: AgentMessage | None = None
    tool_results: list[ToolResultMessage] = field(default_factory=list)


@dataclass
class MessageStartEvent:
    type: Literal["message_start"] = "message_start"
    message: AgentMessage | None = None


@dataclass
class MessageUpdateEvent:
    type: Literal["message_update"] = "message_update"
    message: AgentMessage | None = None
    assistant_message_event: AssistantMessageEvent | None = None


@dataclass
class MessageEndEvent:
    type: Literal["message_end"] = "message_end"
    message: AgentMessage | None = None


@dataclass
class ToolExecutionStartEvent:
    type: Literal["tool_execution_start"] = "tool_execution_start"
    tool_call_id: str = ""
    tool_name: str = ""
    args: JsonObject | None = None


@dataclass
class ToolExecutionUpdateEvent:
    type: Literal["tool_execution_update"] = "tool_execution_update"
    tool_call_id: str = ""
    tool_name: str = ""
    args: JsonObject | None = None
    partial_result: AgentToolResult | None = None


@dataclass
class ToolExecutionEndEvent:
    type: Literal["tool_execution_end"] = "tool_execution_end"
    tool_call_id: str = ""
    tool_name: str = ""
    result: AgentToolResult | None = None
    is_error: bool = False


AgentEvent = Union[
    AgentStartEvent,
    AgentEndEvent,
    TurnStartEvent,
    TurnEndEvent,
    MessageStartEvent,
    MessageUpdateEvent,
    MessageEndEvent,
    ToolExecutionStartEvent,
    ToolExecutionUpdateEvent,
    ToolExecutionEndEvent,
]


@dataclass
class AgentLoopConfig:
    """Configuration for the agent loop."""

    model: Model
    convert_to_llm: ConvertToLlmFn
    transform_context: TransformContextFn | None = None
    get_api_key: ApiKeyResolver | None = None
    get_steering_messages: AgentMessageProvider | None = None
    get_follow_up_messages: AgentMessageProvider | None = None
    api_key: str | None = None
    temperature: float | None = None
    max_tokens: int | None = None


class AgentState(_AgentBaseModel):
    """Agent state containing all configuration and conversation data."""

    system_prompt: str = ""
    model: Model | None = None
    thinking_level: ThinkingLevel = ThinkingLevel.OFF
    tools: list[AgentTool] = Field(default_factory=list)
    messages: list[AgentMessage] = Field(default_factory=list)
    is_streaming: bool = False
    stream_message: AgentMessage | None = None
    pending_tool_calls: set[str] = Field(default_factory=set)
    error: str | None = None


class EventStream:
    """Async event stream that yields events and returns a final result.

    Important: Some agent loops run in background tasks via `asyncio.create_task()`.
    If that background task fails, we must propagate the exception to the consumer
    of the stream. Otherwise callers awaiting `agent.prompt()` can hang forever.
    """

    _WAKEUP_SENTINEL = object()

    def __init__(
        self,
        is_end_event: Callable[[AgentEvent], bool],
        get_result: Callable[[AgentEvent], list[AgentMessage]],
    ):
        self._queue: asyncio.Queue[AgentEvent | object] = asyncio.Queue()
        self._is_end_event = is_end_event
        self._get_result = get_result
        self._result: list[AgentMessage] | None = None
        self._ended = False
        self._exception: BaseException | None = None

    def push(self, event: AgentEvent) -> None:
        if self._ended:
            return
        self._queue.put_nowait(event)

    def end(self, result: list[AgentMessage]) -> None:
        self._result = result
        self._ended = True
        self._queue.put_nowait(self._WAKEUP_SENTINEL)

    def set_exception(self, exc: BaseException) -> None:
        """Terminate the stream with an exception.

        The next consumer read will raise `exc` once all already-queued events
        are drained.
        """

        if self._ended:
            return
        self._exception = exc
        self._ended = True
        self._queue.put_nowait(self._WAKEUP_SENTINEL)

    def __aiter__(self) -> AsyncIterator[AgentEvent]:
        return self

    async def __anext__(self) -> AgentEvent:
        while True:
            if self._queue.empty():
                if self._exception is not None:
                    exc = self._exception
                    self._exception = None
                    raise exc
                if self._ended:
                    raise StopAsyncIteration

            queued_item = await self._queue.get()
            if queued_item is self._WAKEUP_SENTINEL:
                if self._exception is not None:
                    exc = self._exception
                    self._exception = None
                    raise exc
                if self._ended:
                    raise StopAsyncIteration
                continue

            event = cast(AgentEvent, queued_item)
            if self._is_end_event(event):
                self._result = self._get_result(event)
                self._ended = True
            return event

    async def result(self) -> list[AgentMessage]:
        while True:
            if self._queue.empty() and self._exception is not None:
                exc = self._exception
                self._exception = None
                raise exc
            if self._ended and self._queue.empty():
                break
            try:
                await self.__anext__()
            except StopAsyncIteration:
                break

        return self._result or []

File: tinyagent-rs/src/agent_types.rs
use crate::error::Result;
use futures::future::BoxFuture;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use std::collections::{BTreeSet, HashMap};
use std::fmt;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};
use tokio_util::sync::CancellationToken;

pub type JsonObject = Map<String, Value>;
pub type ToolUpdateCallback = Arc<dyn Fn(AgentToolResult) + Send + Sync>;
pub type ToolExecutor = Arc<
    dyn Fn(
            String,
            JsonObject,
            CancellationToken,
            ToolUpdateCallback,
        ) -> BoxFuture<'static, Result<AgentToolResult>>
        + Send
        + Sync,
>;

pub fn current_timestamp() -> i64 {
    match SystemTime::now().duration_since(UNIX_EPOCH) {
        Ok(duration) => duration.as_millis() as i64,
        Err(error) => -(error.duration().as_millis() as i64),
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum ThinkingLevel {
    Off,
    Minimal,
    Low,
    Medium,
    High,
    Xhigh,
}

impl Default for ThinkingLevel {
    fn default() -> Self {
        Self::Off
    }
}

impl ThinkingLevel {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Off => "off",
            Self::Minimal => "minimal",
            Self::Low => "low",
            Self::Medium => "medium",
            Self::High => "high",
            Self::Xhigh => "xhigh",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum QueueMode {
    OneAtATime,
    All,
}

impl Default for QueueMode {
    fn default() -> Self {
        Self::OneAtATime
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub struct ThinkingBudgets {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub thinking_budget: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub struct TextContent {
    #[serde(rename = "type", default = "default_text_kind")]
    pub kind: String,
    #[serde(default)]
    pub text: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub text_signature: Option<String>,
}

fn default_text_kind() -> String {
    "text".to_string()
}

impl TextContent {
    pub fn new(text: impl Into<String>) -> Self {
        Self {
            kind: default_text_kind(),
            text: text.into(),
            text_signature: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub struct ImageContent {
    #[serde(rename = "type", default = "default_image_kind")]
    pub kind: String,
    pub url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
}

fn default_image_kind() -> String {
    "image".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub struct ThinkingContent {
    #[serde(rename = "type", default = "default_thinking_kind")]
    pub kind: String,
    #[serde(default)]
    pub thinking: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub thinking_signature: Option<String>,
}

fn default_thinking_kind() -> String {
    "thinking".to_string()
}

impl ThinkingContent {
    pub fn new(thinking: impl Into<String>) -> Self {
        Self {
            kind: default_thinking_kind(),
            thinking: thinking.into(),
            thinking_signature: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct ToolCallContent {
    #[serde(rename = "type", default = "default_tool_call_kind")]
    pub kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default)]
    pub arguments: Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub partial_json: Option<String>,
}

fn default_tool_call_kind() -> String {
    "tool_call".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum UserContent {
    Text(TextContent),
    Image(ImageContent),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum AssistantContent {
    Text(TextContent),
    Thinking(ThinkingContent),
    ToolCall(ToolCallContent),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub struct Model {
    pub provider: String,
    pub id: String,
    pub api: String,
    #[serde(default)]
    pub thinking_level: ThinkingLevel,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub base_url: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub headers: Option<HashMap<String, String>>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum StopReason {
    Complete,
    Error,
    Aborted,
    ToolCalls,
    Stop,
    Length,
    ToolUse,
}

impl StopReason {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Complete => "complete",
            Self::Error => "error",
            Self::Aborted => "aborted",
            Self::ToolCalls => "tool_calls",
            Self::Stop => "stop",
            Self::Length => "length",
            Self::ToolUse => "tool_use",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct UserMessage {
    pub role: String,
    #[serde(default)]
    pub content: Vec<UserContent>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<i64>,
}

impl Default for UserMessage {
    fn default() -> Self {
        Self {
            role: "user".to_string(),
            content: Vec::new(),
            timestamp: Some(current_timestamp()),
        }
    }
}

impl UserMessage {
    pub fn text(text: impl Into<String>) -> Self {
        Self {
            content: vec![UserContent::Text(TextContent::new(text))],
            ..Self::default()
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AssistantMessage {
    pub role: String,
    #[serde(default)]
    pub content: Vec<Option<AssistantContent>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub stop_reason: Option<StopReason>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub api: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub usage: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,
}

impl Default for AssistantMessage {
    fn default() -> Self {
        Self {
            role: "assistant".to_string(),
            content: Vec::new(),
            stop_reason: None,
            timestamp: Some(current_timestamp()),
            api: None,
            provider: None,
            model: None,
            usage: None,
            error_message: None,
        }
    }
}

impl AssistantMessage {
    pub fn push_content(&mut self, index: usize, content: AssistantContent) {
        if self.content.len() <= index {
            self.content.resize(index + 1, None);
        }
        self.content[index] = Some(content);
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolResultMessage {
    pub role: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_name: Option<String>,
    #[serde(default)]
    pub content: Vec<UserContent>,
    #[serde(default)]
    pub details: Value,
    #[serde(default)]
    pub is_error: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<i64>,
}

impl Default for ToolResultMessage {
    fn default() -> Self {
        Self {
            role: "tool_result".to_string(),
            tool_call_id: None,
            tool_name: None,
            content: Vec::new(),
            details: Value::Object(JsonObject::new()),
            is_error: false,
            timestamp: Some(current_timestamp()),
        }
    }
}

impl ToolResultMessage {
    pub fn from_result(
        tool_call_id: impl Into<String>,
        tool_name: impl Into<String>,
        result: AgentToolResult,
        is_error: bool,
    ) -> Self {
        Self {
            tool_call_id: Some(tool_call_id.into()),
            tool_name: Some(tool_name.into()),
            content: result.content,
            details: result.details,
            is_error,
            ..Self::default()
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct CustomAgentMessage {
    pub role: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payload: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum Message {
    User(UserMessage),
    Assistant(AssistantMessage),
    ToolResult(ToolResultMessage),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum AgentMessage {
    User(UserMessage),
    Assistant(AssistantMessage),
    ToolResult(ToolResultMessage),
    Custom(CustomAgentMessage),
}

impl AgentMessage {
    pub fn role(&self) -> &str {
        match self {
            Self::User(message) => &message.role,
            Self::Assistant(message) => &message.role,
            Self::ToolResult(message) => &message.role,
            Self::Custom(message) => &message.role,
        }
    }

    pub fn as_message(&self) -> Option<Message> {
        match self {
            Self::User(message) => Some(Message::User(message.clone())),
            Self::Assistant(message) => Some(Message::Assistant(message.clone())),
            Self::ToolResult(message) => Some(Message::ToolResult(message.clone())),
            Self::Custom(_) => None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct AgentToolResult {
    #[serde(default)]
    pub content: Vec<UserContent>,
    #[serde(default)]
    pub details: Value,
}

#[derive(Clone, Default)]
pub struct AgentTool {
    pub name: String,
    pub description: String,
    pub parameters: Value,
    pub label: String,
    pub(crate) executor: Option<ToolExecutor>,
}

impl fmt::Debug for AgentTool {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("AgentTool")
            .field("name", &self.name)
            .field("description", &self.description)
            .field("parameters", &self.parameters)
            .field("label", &self.label)
            .finish()
    }
}

impl AgentTool {
    pub fn new<F, Fut>(
        name: impl Into<String>,
        description: impl Into<String>,
        parameters: Value,
        execute: F,
    ) -> Self
    where
        F: Fn(String, JsonObject, CancellationToken, ToolUpdateCallback) -> Fut
            + Send
            + Sync
            + 'static,
        Fut: std::future::Future<Output = Result<AgentToolResult>> + Send + 'static,
    {
        Self {
            name: name.into(),
            description: description.into(),
            parameters,
            label: String::new(),
            executor: Some(Arc::new(move |tool_call_id, args, abort, on_update| {
                Box::pin(execute(tool_call_id, args, abort, on_update))
            })),
        }
    }

    pub fn call(
        &self,
        tool_call_id: String,
        args: JsonObject,
        abort: CancellationToken,
        on_update: ToolUpdateCallback,
    ) -> BoxFuture<'static, Result<AgentToolResult>> {
        match &self.executor {
            Some(executor) => executor(tool_call_id, args, abort, on_update),
            None => Box::pin(async { Ok(AgentToolResult::default()) }),
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct Context {
    pub system_prompt: String,
    pub messages: Vec<Message>,
    pub tools: Vec<AgentTool>,
}

#[derive(Debug, Clone, Default)]
pub struct AgentContext {
    pub system_prompt: String,
    pub messages: Vec<AgentMessage>,
    pub tools: Vec<AgentTool>,
}

impl AgentContext {
    pub fn to_llm_context(&self) -> Context {
        let messages = self
            .messages
            .iter()
            .filter_map(AgentMessage::as_message)
            .collect::<Vec<_>>();

        Context {
            system_prompt: self.system_prompt.clone(),
            messages,
            tools: self.tools.clone(),
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct AgentState {
    pub system_prompt: String,
    pub model: Option<Model>,
    pub thinking_level: ThinkingLevel,
    pub tools: Vec<AgentTool>,
    pub messages: Vec<AgentMessage>,
    pub is_streaming: bool,
    pub stream_message: Option<AgentMessage>,
    pub pending_tool_calls: BTreeSet<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AssistantMessageEvent {
    Start {
        partial: AssistantMessage,
    },
    TextStart {
        content_index: usize,
        partial: AssistantMessage,
    },
    TextDelta {
        content_index: usize,
        delta: String,
        partial: AssistantMessage,
    },
    TextEnd {
        content_index: usize,
        content: String,
        partial: AssistantMessage,
    },
    ThinkingStart {
        content_index: usize,
        partial: AssistantMessage,
    },
    ThinkingDelta {
        content_index: usize,
        delta: String,
        partial: AssistantMessage,
    },
    ThinkingEnd {
        content_index: usize,
        content: String,
        partial: AssistantMessage,
    },
    ToolCallStart {
        content_index: usize,
        partial: AssistantMessage,
    },
    ToolCallDelta {
        content_index: usize,
        delta: String,
        partial: AssistantMessage,
    },
    ToolCallEnd {
        content_index: usize,
        tool_call: ToolCallContent,
        partial: AssistantMessage,
    },
    Done {
        reason: StopReason,
        message: AssistantMessage,
    },
    Error {
        reason: StopReason,
        error: AssistantMessage,
    },
}

impl AssistantMessageEvent {
    pub fn partial(&self) -> Option<&AssistantMessage> {
        match self {
            Self::Start { partial }
            | Self::TextStart { partial, .. }
            | Self::TextDelta { partial, .. }
            | Self::TextEnd { partial, .. }
            | Self::ThinkingStart { partial, .. }
            | Self::ThinkingDelta { partial, .. }
            | Self::ThinkingEnd { partial, .. }
            | Self::ToolCallStart { partial, .. }
            | Self::ToolCallDelta { partial, .. }
            | Self::ToolCallEnd { partial, .. } => Some(partial),
            Self::Done { message, .. } => Some(message),
            Self::Error { error, .. } => Some(error),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AgentEvent {
    AgentStart,
    AgentEnd {
        messages: Vec<AgentMessage>,
    },
    TurnStart,
    TurnEnd {
        message: Option<AgentMessage>,
        tool_results: Vec<ToolResultMessage>,
    },
    MessageStart {
        message: Option<AgentMessage>,
    },
    MessageUpdate {
        message: Option<AgentMessage>,
        assistant_message_event: Option<AssistantMessageEvent>,
    },
    MessageEnd {
        message: Option<AgentMessage>,
    },
    ToolExecutionStart {
        tool_call_id: String,
        tool_name: String,
        args: Option<Value>,
    },
    ToolExecutionUpdate {
        tool_call_id: String,
        tool_name: String,
        args: Option<Value>,
        partial_result: Option<AgentToolResult>,
    },
    ToolExecutionEnd {
        tool_call_id: String,
        tool_name: String,
        result: Option<AgentToolResult>,
        is_error: bool,
    },
}

impl AgentEvent {
    pub fn kind(&self) -> &'static str {
        match self {
            Self::AgentStart => "agent_start",
            Self::AgentEnd { .. } => "agent_end",
            Self::TurnStart => "turn_start",
            Self::TurnEnd { .. } => "turn_end",
            Self::MessageStart { .. } => "message_start",
            Self::MessageUpdate { .. } => "message_update",
            Self::MessageEnd { .. } => "message_end",
            Self::ToolExecutionStart { .. } => "tool_execution_start",
            Self::ToolExecutionUpdate { .. } => "tool_execution_update",
            Self::ToolExecutionEnd { .. } => "tool_execution_end",
        }
    }
}

pub fn extract_text(message: &AgentMessage) -> String {
    match message {
        AgentMessage::User(message) => message
            .content
            .iter()
            .filter_map(|content| match content {
                UserContent::Text(text) => Some(text.text.as_str()),
                UserContent::Image(_) => None,
            })
            .collect(),
        AgentMessage::Assistant(message) => message
            .content
            .iter()
            .filter_map(|content| match content {
                Some(AssistantContent::Text(text)) => Some(text.text.as_str()),
                _ => None,
            })
            .collect(),
        AgentMessage::ToolResult(message) => message
            .content
            .iter()
            .filter_map(|content| match content {
                UserContent::Text(text) => Some(text.text.as_str()),
                UserContent::Image(_) => None,
            })
            .collect(),
        AgentMessage::Custom(_) => String::new(),
    }
}
